controllers/stardoginstance_controller.go

```diff
@@ -7,36 +7,221 @@ package controllers
 // +kubebuilder:rbac:groups=stardog.vshn.ch,resources=stardoginstances/status,verbs=get;update;patch
 
 func (r *StardogInstanceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("stardoginstance", req.NamespacedName)
+	namespace := req.NamespacedName
+	_ = r.Log.WithValues("stardoginstance", namespace)
+
+	ctx := context.Background()
+	stardogInstance := &StardogInstance{}
+
+	err := r.Client.Get(ctx, namespace, stardogInstance)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info("StardogInstance not found, ignoring reconcile.", "StardogInstance", namespace)
+			return ctrl.Result{Requeue: false}, nil
+		}
+		r.Log.Error(err, "could not retrieve StardogInstance.", "StardogInstance", namespace)
+		return ctrl.Result{Requeue: true, RequeueAfter: r.ReconcileInterval}, err
+	}
+
+	rc := &ReconciliationContext{
+		context:       ctx,
+		conditions:    make(map[StardogConditionType]StardogCondition),
+		namespace:     namespace.Namespace,
+		stardogClient: stardogrest.NewExtendedBaseClient(),
+	}
+	sir := &StardogInstanceReconciliation{
+		reconciliationContext: rc,
+		resource:              stardogInstance,
+	}
+
+	r.Log.Info("reconciling", getLoggingKeysAndValuesForStardogInstance(stardogInstance)...)
+
+	isStardogInstanceMarkedToBeDeleted := stardogInstance.GetDeletionTimestamp() != nil
+	if isStardogInstanceMarkedToBeDeleted {
+		r.Log.Info("checking if StardogInstance is deletable")
+		if r.deleteStardogInstance(sir) != nil {
+			rc.SetStatusCondition(createStatusConditionTerminating(err))
+			rc.SetStatusCondition(createStatusConditionReady(false, "StardogInstance not ready"))
+			return ctrl.Result{Requeue: true}, r.updateStatus(sir)
+		}
+		return ctrl.Result{Requeue: false}, r.updateStatus(sir)
+	}
+
+	if r.validateSpecification(sir.resource.Spec) != nil {
+		rc.SetStatusCondition(createStatusConditionInvalid(err))
+		rc.SetStatusCondition(createStatusConditionReady(false, "StardogInstance not ready"))
+		return ctrl.Result{}, r.updateStatus(sir)
+	}
+	rc.SetStatusIfExisting(StardogInvalid, v1.ConditionFalse)
 
-	// your logic here
+	if r.validateConnection(sir) != nil {
+		rc.SetStatusCondition(createStatusConditionErrored(err))
+		rc.SetStatusCondition(createStatusConditionReady(false, "StardogInstance not ready"))
+		return ctrl.Result{}, r.updateStatus(sir)
+	}
+	rc.SetStatusIfExisting(StardogErrored, v1.ConditionFalse)
 
-	return ctrl.Result{}, nil
+	if missingAtLeastOne(sir.resource.GetFinalizers(), instanceUserFinalizer, instanceRoleFinalizer) {
+		r.Log.V(1).Info(fmt.Sprintf("adding %s for the StardogInstance", instanceUserFinalizer))
+		controllerutil.AddFinalizer(sir.resource, instanceUserFinalizer)
+		r.Log.V(1).Info(fmt.Sprintf("adding %s for the StardogInstance", instanceRoleFinalizer))
+		controllerutil.AddFinalizer(sir.resource, instanceRoleFinalizer)
+	}
```

Comment:
The AddFinalizer has already a `if-not-present` condition built-in, so no need to check ourselves whether it's missing.
```suggestion
	controllerutil.AddFinalizer(sir.resource, instanceUserFinalizer)
	controllerutil.AddFinalizer(sir.resource, instanceRoleFinalizer)
```